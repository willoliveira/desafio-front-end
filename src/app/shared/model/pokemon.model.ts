import { IRessource } from "./ressource.model";

export enum EnumPokemonTypes {
	'normal',
	'fighting',
	'flying',
	'poison',
	'ground',
	'rock',
	'bug',
	'ghost',
	'steel',
	'fire',
	'water',
	'grass',
	'electric',
	'psychic',
	'ice',
	'dragon',
	'dark',
	'fairy',
	'unknown',
	'shadow'
}

export interface IPokemonAbilities {
	ability: IRessource;
	is_hidden: boolean;
	slot: number;
}

export interface IPokemonStats {
	base_stat: number;
	effort: number;
	stat: IRessource;
}

export interface IPokemonTypes {
	slot: number;
	type: IRessource;
}

export interface IPokemonSprites {
	back_default: string;
	back_female: string;
	back_shiny: string;
	back_shiny_female: string;
	front_default: string;
	front_female: string;
	front_shiny: string;
	front_shiny_female: string;
}

export interface IPokemon {
	id: number;
	base_experience: number;
	name: string;
	sprites: IPokemonSprites;
	weight: number;
	height: number;
	types: IPokemonTypes[];
	abilities: IPokemonAbilities[];
	stats: IPokemonStats[];
}
