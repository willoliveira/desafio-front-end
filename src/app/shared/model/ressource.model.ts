export interface IRessource {
	name: string;
	url: string;
}
