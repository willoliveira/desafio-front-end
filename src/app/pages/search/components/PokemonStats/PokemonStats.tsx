import React, { FC } from 'react';
import { IPokemonStats } from '../../../../shared/model/pokemon.model';
import { View, StyleSheet, Text } from 'react-native';

export interface IPokemonStatsProps {
	stats: IPokemonStats[];
}

export const PokemonStats: FC<IPokemonStatsProps> = ({ stats }: IPokemonStatsProps) => (
	<View style={styles.statsContainer}>
		<Text>Stats</Text>
		<View style={styles.statsParent}>
			{
				stats.map((stat: IPokemonStats) => (
					<View key={`key-${stat.stat.name}`} style={styles.bar}>
						<View style={styles.statsBar} />
						<View style={{ ...styles.statsFillBar, height: `${calcStatsFillPercent(stat)}%` }}>
							<Text style={styles.statsFillText}>{stat.base_stat}</Text>
						</View>
						<Text style={styles.statsName}>{statsTranspile[stat.stat.name]}</Text>
					</View>
				))
			}
		</View>
	</View>
);

export const calcStatsFillPercent = (stat: IPokemonStats, base_stats: number = 200): number => {
	return Math.round((stat.base_stat / base_stats) * 100);
}

export const statsTranspile = {
	'speed': 'Spd',
	'special-defense': 'Sp Def',
	'special-attack': 'Sp Atk',
	'defense': 'Def',
	'attack': 'Atk',
	'hp': 'HP'
}


const styles = StyleSheet.create({
	statsContainer: {
		backgroundColor: '#a4a4a4',
		borderRadius: 10,
		minHeight: 250,
		width: '100%',
		paddingTop: 18,
		paddingBottom: 42,
		paddingLeft: 12,
		paddingRight: 12,
		marginBottom: 12,
		flex: 1,
		flexDirection: "column"
	},
	statsParent: {
		display: "flex",
		flex: 1,
		flexDirection: 'row',
		justifyContent: 'space-between',
		height: '90%'
	},
	bar: {
		height: '100%',
		flex: 1,
		flexGrow: 1,
		margin: 2
	},
	statsBar: {
		backgroundColor: '#FFFFFF',
		position: 'absolute',
		height: '100%',
		width: '100%'
	},
	statsFillBar: {
		backgroundColor: '#30a7d7',
		position: 'absolute',
		zIndex: 999,
		width: '100%',
		bottom: 0
	},
	statsFillText: {
		position: 'absolute',
		color: '#FFFFFF',
		textAlign: 'center',
		bottom: 4,
		left: 0,
		width: '100%'
	},
	statsName: {
		position: 'absolute',
		bottom: -20,
		textAlign: 'center',
		width: '100%',
		fontWeight: 'bold'
	}
});
