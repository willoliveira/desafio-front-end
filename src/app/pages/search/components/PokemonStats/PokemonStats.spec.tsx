/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import { PokemonStats, calcStatsFillPercent, statsTranspile } from './PokemonStats';
import { IPokemonStats } from '../../../../shared/model/pokemon.model';

describe('<PokemonStats />', () => {
	let stats: IPokemonStats[];
	beforeAll(() => {
		stats = [
			{
				"base_stat": 100,
				"effort": 0,
				"stat": { "name": "speed", "url": "https://pokeapi.co/api/v2/stat/6/" }
			},
			{
				"base_stat": 65,
				"effort": 0,
				"stat": { "name": "special-defense", "url": "https://pokeapi.co/api/v2/stat/5/" }
			},
			{
				"base_stat": 65,
				"effort": 1,
				"stat": { "name": "special-attack", "url": "https://pokeapi.co/api/v2/stat/4/"
				}
			},
			{
				"base_stat": 49,
				"effort": 0,
				"stat": { "name": "defense", "url": "https://pokeapi.co/api/v2/stat/3/" }
			},
			{
				"base_stat": 49,
				"effort": 0,
				"stat": { "name": "attack", "url": "https://pokeapi.co/api/v2/stat/2/" }
			},
			{
				"base_stat": 45,
				"effort": 0,
				"stat": { "name": "hp", "url": "https://pokeapi.co/api/v2/stat/1/" }
			}
		]
	});

	describe('Rendering', () => {

		it('should match to snapshot - one type - electric', () => {
			const component = shallow(<PokemonStats stats={stats} />);
			expect(component).toMatchSnapshot("PokemonStats snapshot");
		});
	});

	describe('calcStatsFillPercent()', () => {

		it('Calc stats fill bar', () => {
			expect(calcStatsFillPercent(stats[0])).toEqual(50)
		});
	});

	describe('stats transpile', () => {

		it('speed => Spd', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});

		it('special-defense => Sp Def', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});

		it('special-attack => Sp Atk', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});

		it('defense => Def', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});

		it('attack => Atk', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});

		it('hp => HP', () => {
			expect(statsTranspile['speed']).toEqual('Spd')
		});
	});
});
