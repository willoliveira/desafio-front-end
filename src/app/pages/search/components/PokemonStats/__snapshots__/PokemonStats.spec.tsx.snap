// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonStats /> Rendering should match to snapshot - one type - electric: PokemonStats snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#a4a4a4",
      "borderRadius": 10,
      "flex": 1,
      "flexDirection": "column",
      "marginBottom": 12,
      "minHeight": 250,
      "paddingBottom": 42,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 18,
      "width": "100%",
    }
  }
>
  <Text>
    Stats
  </Text>
  <View
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
        "height": "90%",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      key="key-speed"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "50%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          100
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Spd
      </Text>
    </View>
    <View
      key="key-special-defense"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "33%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          65
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Sp Def
      </Text>
    </View>
    <View
      key="key-special-attack"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "33%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          65
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Sp Atk
      </Text>
    </View>
    <View
      key="key-defense"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "25%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          49
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Def
      </Text>
    </View>
    <View
      key="key-attack"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "25%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          49
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Atk
      </Text>
    </View>
    <View
      key="key-hp"
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "height": "100%",
          "margin": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#30a7d7",
            "bottom": 0,
            "height": "23%",
            "position": "absolute",
            "width": "100%",
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "bottom": 4,
              "color": "#FFFFFF",
              "left": 0,
              "position": "absolute",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          45
        </Text>
      </View>
      <Text
        style={
          Object {
            "bottom": -20,
            "fontWeight": "bold",
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        HP
      </Text>
    </View>
  </View>
</View>
`;
